Remove multiprocess finalizer and improve subprocess docs (#265)
pytest-dev/pytest-cov: docs/subprocess-support.rst

==================
Subprocess support
==================

> Although pytest-cov supports subprocesses and multiprocessing. However, there are few pitfalls that need to be

Reviewer: ```suggestion
pytest-cov supports subprocesses and multiprocessing. However, there are a few pitfalls that need to be
```

explained.

> Normally coverage writes the data via a pretty standard atexit handler. However, if the subprocess doesn't exit on it's

Reviewer: ```suggestion
Normally coverage writes the data via a pretty standard atexit handler. However, if the subprocess doesn't exit on its
```

own then the atexit handler might not run. Why that happens is best left to the adventurous to discover by waddling

> though the python bug tracker.

Reviewer: ```suggestion
though the Python bug tracker.
```

        finally:       # <= THIS IS ESSENTIAL
            p.close()  # <= THIS IS ESSENTIAL
            p.join()   # <= THIS IS ESSENTIAL

Previously this guide recommended using ``multiprocessing.Pool``'s context manager API, however, that was wrong as
``multiprocessing.Pool.__exit__`` is an alias to ``multiprocessing.Pool.terminate``, and that doesn't always run the
finalizers (sometimes the problem in `cleanup_on_sigterm`_ will appear).

If you use ``multiprocessing.Process``
======================================

There's an identical issue when using the ``Process`` objects. Don't forget to use ``.join()``:

.. code-block:: python

    from multiprocessing import Process

    def f(name):
        print('hello', name)

    if __name__ == '__main__':
        p = Process(target=f, args=('bob',))
        try:
            p.start()
        finally:      # <= THIS IS ESSENTIAL
            p.join()  # <= THIS IS ESSENTIAL

.. _cleanup_on_sigterm:

If you abuse ``multiprocessing.Process.terminate``
==================================================

It appears that many people are using the ``terminate`` method and then get unreliable coverage results.

> On Linux usually that means a SIGTERM gets sent to the process. Unfortunately Python don't have a default handler for

Reviewer: ```suggestion
On Linux usually that means a SIGTERM gets sent to the process. Unfortunately Python doesn't have a default handler for
```

SIGTERM so you need to install your own. Because ``pytest-cov`` doesn't want to second-guess (not yet, add your thoughts
on the issue tracker if you disagree) it doesn't install a handler by default, but you can activate it by doing this:

.. code-block:: python

    try:
        from pytest_cov.embed import cleanup_on_sigterm
    except ImportError:
        pass
    else:
        cleanup_on_sigterm()

> On Windows there's no nice way to do cleanup (no signal handlers) so you're left to your own devices.

Reviewer: ```suggestion
On Windows there's no nice way to do cleanup (no signal handlers).
```

If anything else
================

> If you have custom signal handling, eg: you do reload on SIGHUP you should have something like this:

Reviewer: ```suggestion
If you have custom signal handling, e.g. you do reload on SIGHUP you should have something like this:
```
